### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'contains')

### ISSUE THREAD

**Issue by reintroducing** [focus-trap/focus-trap-react#905]
We're using Sentry to log errors in our production application and I randomly see this error pop up from time to time. I haven't been able to figure out why it is happening so I've mostly ignored it, but I actually was able to reproduce it locally today by complete accident. I have a custom Modal component that uses `focus-trap-react` and I believe that is where the error originates. I'm using the latest version of Chrome on the latest version of MacOS so I don't think this is specific to any certain browser (I've seen the Sentry error come through in Chrome on Windows 10 as well from our production app users).

This is the error I'm seeing:
![Screenshot 2023-01-24 at 8 29 15 AM](https://user-images.githubusercontent.com/259901/214336202-5e0750f3-d713-4567-8281-e24c224ec5c7.png)
It is interesting how/why its happening. For this particular case, I have the modal open and all I'm doing is automatically calling a download link for a PDF (and a dummy PDF at that) as the modal is closing:
```js
const onDownloadPdf = () => {
    const a = document.createElement('a');

    a.href =
        'https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf';
    a.target = '_blank';

    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
};
```
While the download is happening and opening a new tab with the PDF, the modal is closing in the other tab at the same time. This seems to be an important distinction because if I comment out the code inside of `onDownloadPdf` but still call the method, it does not error. It's almost as if the concept of opening a new tab with a PDF in it is causing this error, but I can't figure out the correlation nor track it down further (it just doesn't make much sense to me as to why this would be happening).

I'm posting here in hopes of possibly triggering something in your mind, since you know the internals of the codebase, that may register as a possible issue based on the above information so that hopefully we can get to the bottom of this. I know this is pretty vague but perhaps it may be helpful in trying to resolve this.

**Comment by stefcameron** [comment]
@reintroducing Thank you for the information you were able to provide, and the callstack. Assuming the stack is not misleading (sometimes source maps, etc., confuse things), and since the error is related to accessing a `contains` property off something that's null/undefined, and given your particular use case of attaching a node to the document, clicking it (maybe even a programmatic click causes a focus change?), and then immediately removing it from the document, I'm guessing the error is coming from here: https://github.com/focus-trap/tabbable/blob/master/src/index.js#L277

```javascript
  let attached = !!(
    nodeRootHost?.ownerDocument.contains(nodeRootHost) ||      // <- HERE
    node.ownerDocument.contains(node)
  );
```

Based on the [ownerDocument docs](https://developer.mozilla.org/en-US/docs/Web/API/Node/ownerDocument), there's a case where this property could return `null`, that being when it's accessed on a document -- something the code is currently _not_ protecting against (and I think _should_).

And then there's the fact that you're immediately detaching the node. If that node somehow got focus for a split second and then was removed, it's possible focus was diverted back to the document (happens sometimes), which could lead to this scenario where we're checking something essentially like `document.ownerDocument` and that results in `null`. Then we access `contains` on that and 💥 .

But I'm just guessing... If I post a beta of `tabbable` with this fix, would you have a way to test it to see if I'm right? I'm not sure if I would also have to post a beta of `focus-trap` and `focus-trap-react` or if there's a way to override the version of tabbable referenced by focus-trap, referenced by focus-trap-react in the lock file just for a one-off test.

**Comment by reintroducing** [comment]
@stefcameron I'm glad you were able to track this down and your explanation makes perfect sense. I'm not sure how I can beta just tabbable since its a nested dependency, but what we can possibly do is just drop a new `dist` directory into `node_modules/tabbable` and maybe that will recognize the changes and allow me to test them properly without having to mess with multiple versions of packages? manually doing this is not ideal, especially since vite sometimes has the tendency to cache dependencies and it seems like no matter what i try i can't bust them to test local changes to my node_modules, but we could try that first if publishing 3 betas of these packages is too much of a headache for you. 

There's also the added complexity of the fact that these dependencies are in my UI library which in itself is a separate repo that gets published to a private npm registry, so that makes it even more fun :)

**Comment by stefcameron** [comment]
@reintroducing Oh how fun dependency land is! 😆 OK, I think I'll just publish a one-off of the whole stack so that we're more certain whether it works or not is because it was fixed or not, instead of wondering if it's not fixed because the old dependency is still being taken. I'll ping you again here once I've got that done.

**Comment by reintroducing** [comment]
@stefcameron sounds good, just let me know when ready and I’ll give them a spin

**Comment by stefcameron** [comment]
@reintroducing OK, this isn't elegant, but hopefully using `focus-trap-react@10.0.3-tabbable-i905.0` will pull in everything down to `tabbable@6.0.2-tabbable-i905.0` and allow us to see if my assumptions about the crash were correct. Please give it a try!

**Comment by reintroducing** [comment]
@stefcameron just gave that a shot but unfortunately it produced a new error:
![Screenshot 2023-01-25 at 1 47 46 PM](https://user-images.githubusercontent.com/259901/214687082-44af8a4c-c658-4101-8fe6-408df617a4bd.png)

**Comment by stefcameron** [comment]
@reintroducing Thanks for trying it. Please try `focus-trap-react@10.0.3-tabbable-i905.1` (`1` at the end has changed) and LMK.

**Comment by reintroducing** [comment]
@stefcameron hmm, that does not seem to have changed anything, i'm still getting the same exact error.

**Comment by stefcameron** [comment]
@reintroducing Thanks for trying it out. 😞 I think I'm going to need a small, isolated repro if I'm going to figure this one out. I'm really shooting blind and this feels like a real corner case. Are you able to provide me with a sandbox somewhere? Or simple code here I can copy paste into our [demo](https://github.com/focus-trap/focus-trap-react/tree/master/demo) environment (which is ultimately what I'll need to do to play with it and debug it)?

**Comment by reintroducing** [comment]
@stefcameron yeah, I stopped short of saying I’ll try to make a small repro for you earlier because it was end of the work day, but I’m going to try and see if I can extract the pertinent code from our UI library and then the stuff using it in the project that reproduces this and boil it down to as basic as possible. I realize this is a giant edge case and as such not easy to solve so let me see if I can put something together. Regardless I really appreciate you taking the time to look into this and work through it with me.

**Comment by stefcameron** [comment]
> @stefcameron yeah, I stopped short of saying I’ll try to make a small repro for you earlier because it was end of the work day, but I’m going to try and see if I can extract the pertinent code from our UI library and then the stuff using it in the project that reproduces this and boil it down to as basic as possible. I realize this is a giant edge case and as such not easy to solve so let me see if I can put something together. Regardless I really appreciate you taking the time to look into this and work through it with me.

My pleasure!

**Comment by reintroducing** [comment]
@stefcameron I was able to put together an example of the issue and am attaching a zip.

[focus-trap-react-bug.zip](https://github.com/focus-trap/focus-trap-react/files/10511185/focus-trap-react-bug.zip)

If you grab that, run `npm i` and then `npm start`, you'll see an Open button. Click that, then either hit close in the modal or click on the darker shim part and itll close, open a new window with a PDF, close that, and you'll see the error in the console for focus trap. 

I tried to clean this up as much as possible from code in two repos so apologies if its still a bit messy but it does reproduce the issue correctly. Let me know if you have any questions.

**Comment by stefcameron** [comment]
@reintroducing Thank you for that repro, it was really helpful! I found the issue right away, that being that in some cases, a detached node's root node will be the node itself instead of `null`. This is the case with the anchor that's in the DOM for all of ~300ms, and which your code clicks -- still long enough for focus-trap to register a change in focus and then check if the node is focusable, and... crash.

Version `10.0.3-tabbable-i905.2` of focus-trap-react should fix the problem. Could you confirm?

**Comment by reintroducing** [comment]
@stefcameron awesome, i'm glad that it was helpful! i just gave that beta a try and can confirm that it is working without issue. thank you so much! this has been such a longstanding issue in Sentry for us that I'll be thrilled to finally get rid of it :)

**Comment by stefcameron** [comment]
Thanks for checking it out! I'm glad we could get to the bottom of it, thanks your help.

The fix is all the down in tabbable, which means I'll have to publish everything again, and there is a small feature I need to add to tabbable, so this fix likely won't land for a couple of weeks yet, but it will land soon for sure.

**Comment by reintroducing** [comment]
No worries, sounds good! luckily it doesn't really break any functionality, its just an annoying error that pops up here and there, so i can definitely wait until you're ready to release. thanks again!

**Comment by stefcameron** [comment]
@reintroducing Finally, the fix is available in v10.1.0!
